File: src/FarmEngine.Core/Skills.cs
using FarmEngine.Json;
using FarmEngine.Schemas;

namespace FarmEngine.Core;

/// <summary>
/// Player skills (M4g): XP per action category; levels unlock recipes and
/// grant small passive modifiers. Toggleable per project. Port of skills.ts.
/// </summary>
public static class Skills
{
    public static double SkillLevelForXp(List<double> curve, double xp)
    {
        double level = 0;
        for (var i = 0; i < curve.Count; i++)
        {
            if (xp >= curve[i]) level = i;
        }
        return level;
    }

    public static EngineStep GrantXp(EngineContext ctx, GameState state, string skill, double xp)
    {
        if (!ctx.Content.Settings.SkillsEnabled || xp <= 0) return EngineStep.Of(state);

        var current = state.Player.Skills.TryGetValue(skill, out var existing) && existing != null
            ? existing
            : new SkillState { Xp = 0, Level = 0 };
        var newXp = current.Xp + xp;
        var newLevel = SkillLevelForXp(ctx.Content.Settings.SkillLevelCurve, newXp);
        var effects = new List<Effect>();
        if (newLevel > current.Level)
        {
            var label = skill.Length == 0 ? "" : skill[..1].ToUpperInvariant() + skill[1..];
            effects.Add(Effect.Message("success", $"{label} level {Js.Num(newLevel)}!"));
        }

        var skills = new OrderedDictionary<string, SkillState>(state.Player.Skills)
        {
            [skill] = new SkillState { Xp = newXp, Level = newLevel },
        };
        return new EngineStep(state with { Player = state.Player with { Skills = skills } }, effects);
    }

    public static double SkillLevel(GameState state, string skill) =>
        state.Player.Skills.TryGetValue(skill, out var s) && s != null ? s.Level : 0;

    /// <summary>Farming levels add bonus yield: +1 per 4 levels.</summary>
    public static double FarmingYieldBonus(GameState state) =>
        Math.Floor(SkillLevel(state, "farming") / 4);
}

File: crates/farm-sim/src/skills.rs
//! Skills (port of `Skills.cs` / skills.ts).

use crate::engine_types::{Effects, EngineContext};
use crate::schema::GameState;

pub fn skill_level_for_xp(curve: &[f64], xp: f64) -> f64 {
    let _ = (curve, xp);
    todo!("port Skills.SkillLevelForXp")
}

pub fn grant_xp(ctx: &EngineContext, state: &mut GameState, skill: &str, xp: f64) -> Effects {
    let _ = (ctx, state, skill, xp);
    todo!("port Skills.GrantXp")
}

pub fn skill_level(state: &GameState, skill: &str) -> f64 {
    let _ = (state, skill);
    todo!("port Skills.SkillLevel")
}

pub fn farming_yield_bonus(state: &GameState) -> f64 {
    let _ = state;
    todo!("port Skills.FarmingYieldBonus")
}
